crates/sc-consensus-subspace/src/archiver.rs

```diff
@@ -815,29 +809,49 @@ where
     Ok(async move {
         let archiver = match maybe_archiver {
             Some(archiver) => archiver,
-            None => initialize_archiver(&segment_headers_store, &subspace_link, client.as_ref())?,
+            None => {
+                let Some(last_segment_header) = segment_headers_store.last_segment_header() else {
+                    unreachable!("Already checked above that max_segment_index is not None")
+                };
+                let last_archived_block_number =
+                    last_segment_header.last_archived_block().number.into();
+                let mut block_number_to_archive = Zero::zero();
+                // wait for unknown blocks to arrive.
+                while let Some(ref block_import_notification) =
+                    block_importing_notification_stream.next().await
+                {
+                    let block_number = block_import_notification.block_number;
+                    if block_number < last_archived_block_number {
+                        debug!(%block_number, %last_archived_block_number, "skip already know block");
+                        continue;
+                    }
+                    if block_number == last_archived_block_number {
+                        debug!(%block_number, "hit an edge of the last archived block");
+                        let best_block_hash = client.info().best_hash;
+                        block_number_to_archive = last_archived_block_number
+                            + client
+                                .runtime_api()
+                                .chain_constants(best_block_hash)?
+                                .confirmation_depth_k()
+                                .into();
+                        continue;
+                    }
+                    if block_number > block_number_to_archive {
```

Comment:
I'm not sure this condition is correct, the names are definitely off. We archive blocks are `confirmation_depth_k`, if the block being imported is that deep then it is too late already, we've skipped too much. However, according to the code `block_number_to_archive` is not actually the block number to archive.

Reply:
Oh, I think I must have had too much to drink yesterday. I've skipped too much, just initialize the archiver when it reaches the edge. Let me go back to the previous code

Reply:
I opened https://github.com/subspace/subspace/pull/2926 and already have a branch locally that optimizes archiving with basically one line of code

Reply:
Yeah, that's right. Let me close this.

Reply:
Here: https://github.com/subspace/subspace/pull/2927

Reply:
got it